### PR TITLE
review 3.15 spec

### PR DESCRIPTION
**Pull request by aeschli**
Testing https://github.com/microsoft/vscode/issues/81631

There are several TODO that are open questions (don't merge...)

### CODE REVIEW

**Comment by dbaeumer** [comment]
@aeschli thanks. Ping me when ready to merge.

**Comment by dbaeumer** [comment]
@aeschli I can't find any todos in the PR :-). Any pointer?

File: _specifications/specification-3-15.md [hunk 1]
```diff
@@ -4331,6 +4340,8 @@ export interface SymbolInformation {
 * partial result: `DocumentSymbol[]` \| `SymbolInformation[]`
 * error: code and message set in case an exception happens during the document symbol request.
 
+TODO: can partial result be a mix of SymbolInformation and DocumentSymbol ? IMO we could limit it to SymbolInformation
```
Comment (rcjsuen) [inline _specifications/specification-3-15.md]: @dbaeumer Here is one.

File: _specifications/specification-3-15.md [hunk 2]
```diff
@@ -128,10 +128,10 @@ interface ResponseError<D> {
 	message: string;
 
 	/**
-	 * A Primitive or Structured value that contains additional
+	 * A primitive or structured value that contains additional
 	 * information about the error. Can be omitted.
 	 */
-	data?: D;
+	data?: string | number | boolean | number | array | object | null;
```
Comment (h-michael) [inline _specifications/specification-3-15.md]: Sorry if I have misunderstood.
I think `array` is not included TypeScript primitive types.
So I think this is invalid type definition.

http://www.typescriptlang.org/play/?ssl=16&ssc=46&pln=16&pc=51#code/JYOwLgpgTgZghgYwgAgEoQM4AcD2IMQCiUUOUyA3gFACQA9AFQO3IPICCyIArgLYBG0ZKAAmwBHDCgA5sjAALFNFLkwATywoFk5DgQJuJCCIB0LBnVoIcIiAC4ufQVADcVWo2Y1WHZBjBQMshYpABuwGIgsnB+8mRgyLYYCIFYUni6MHKKyMpkZt4WtLyYGHDS9n4BMm4eTOa+IcC8wFKhKGRVUNwIYIbGyKFwADbcWvI61uBwoBjIcCJi6SAjDaAwZLySwBlw-DjcCQpKJPnIAMJwIMiCui1gkKbmljQiknAA-A7+gVHIAD6OARCQH7HDDCBXAFA5zQuAkOBqaE4fgAKwgvWhPGGwzcAF8gA

File: _specifications/specification-3-15.md [hunk 3]
```diff
@@ -165,7 +165,7 @@ interface NotificationMessage extends Message {
 	/**
 	 * The notification's params.
 	 */
-	params?: Array<any> | object;
+	params?: array | object;
```
Comment (h-michael) [inline _specifications/specification-3-15.md]: same as [above](https://github.com/microsoft/language-server-protocol/pull/833/files#r334742968)

File: _specifications/specification-3-15.md (hunk 2 above)
Comment (KamasamaK) [reply _specifications/specification-3-15.md]: Also, `number` is there twice.

File: _specifications/specification-3-15.md [hunk 4]
```diff
@@ -438,13 +438,13 @@ export namespace DiagnosticTag {
      * Clients are allowed to render diagnostics with this tag faded out instead of having
      * an error squiggle.
      */
-    export const Unnecessary: 1;
+    export const Unnecessary = 1;
```
Comment (dbaeumer) [inline _specifications/specification-3-15.md]: This should be export const Unnecessary: 1 = 1;

File: _specifications/specification-3-15.md [hunk 5]
```diff
@@ -438,13 +438,13 @@ export namespace DiagnosticTag {
...
      */
-    export const Unnecessary: 1;
+    export const Unnecessary = 1;
     /**
      * Deprecated or obsolete code.
      *
      * Clients are allowed to rendered diagnostics with this tag strike through.
      */
-    export const Deprecated: 2;
+    export const Deprecated = 2;
```
Comment (dbaeumer) [inline _specifications/specification-3-15.md]: Same as above

File: _specifications/specification-3-15.md (hunk 2 above)
Comment (aeschli) [reply _specifications/specification-3-15.md]: As the LSP is JSON, the valid types must be JSON data types. `any` is more than that.
Maybe we can define:
```ts
type LSPAny = LSPObject | LSPArray | string | number | boolean | null;
type LSPObject = { [key:string]:LSPAny };
type LSPArray = LSPAny[];
```
and use it where needed.

**Comment by Razzeee** [comment]
Are there any concrete plans on when this is targeted for release? Last release was also in December, so will we get a new version every December then? Or how are you deciding when to release?

**Comment by dbaeumer** [comment]
Merged the changes and addressed the open ToDos

**Pull request closed without merging**